ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,907 @@
+		for np in self.notify.read().iter() {
+			if let Some(n) = np.upgrade() {
+				f(&*n);
+			}
+		}
+	}
+
+	/// Register client reference
+	pub fn register_client(&self, client: Weak<Client>) {
+		*self.client.write() = Some(client);
+	}
+
+	/// Register accounts provider
+	pub fn register_account_provider(&self, accounts: Weak<AccountProvider>) {
+		*self.accounts.write() = Some(accounts);
+	}
+
+	/// Sets encryptor
+	pub fn set_encryptor(&self, encryptor: Arc<Encryptor>) {
+		*self.encryptor.write() = encryptor.clone();
+	}
+
+	/// Sets provider's config.
+	pub fn set_config(&self, config: ProviderConfig) -> Result<(), EthcoreError> {
+		let url = config.key_server_url.clone();
+		let threshold = config.key_server_threshold;
+		*self.config.write() = config;
+		//replace encryptor with new one with parameters set
+		*self.encryptor.write() = Arc::new(SecretStoreEncryptor::new(url, threshold)?);
+		Ok(())
+	}
+
+	/// 1. Create private transaction from the signed transaction
+	/// 2. Executes private transaction
+	/// 3. Save it with state returned on prev step to the queue for signing
+	/// 4. Broadcast corresponding message to the chain
+	pub fn create_private_transaction(&self, signed_transaction: SignedTransaction) -> Result<Receipt, EthcoreError> {
+		trace!("Creating private transaction from regular transaction: {:?}", signed_transaction);
+		if self.config.read().signer_account.is_none() {
+			trace!("Signing account not set");
+			return Err(PrivateTransactionError::SignerAccountNotSet.into());
+		}
+		let tx_hash = signed_transaction.hash();
+		match signed_transaction.action {
+			Action::Create => {
+				return Err(PrivateTransactionError::BadTransactonType.into());
+			}
+			Action::Call(contract) => {
+				let data = signed_transaction.rlp_bytes();
+				let encrypted_transaction = self.encrypt(&contract, &Self::iv_from_transaction(&signed_transaction), &data)?;
+				let private = PrivateTransaction {
+					encrypted: encrypted_transaction,
+					contract: contract,
+				};
+				let private_state = self.execute_private_transaction(BlockId::Latest, &signed_transaction)?;
+				trace!("Private transaction created, encrypted transaction: {:?}, private state: {:?}", private, private_state);
+				let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+				trace!("Required validators: {:?}", contract_validators);
+				let private_state_hash = keccak(&private_state);
+				trace!("Hashed effective private state for sender: {:?}", private_state_hash);
+				self.transactions_for_signing.lock().add_transaction(private.hash(), signed_transaction, contract_validators, private_state)?;
+				self.broadcast_private_transaction(private.rlp_bytes().into_vec());
+				Ok(Receipt {
+					hash: tx_hash,
+					contract_address: None,
+					status_code: 0,
+				})
+			}
+		}
+	}
+
+	/// Try to unlock account using stored passwords
+	fn unlock_account(&self, account: &Address) -> Result<bool, EthcoreError> {
```

Comment:
Why do we need account unlocking? It opens up a bunch of race condition attacks.